backend/src/Controllers/finalizeElectionController.ts

```diff
@@ -0,0 +1,56 @@
+import ServiceLocator from '../ServiceLocator';
+import Logger from '../Services/Logging/Logger';
+import { permissions } from '../../../domain_model/permissions';
+import { expectPermission } from "./controllerUtils";
+import { BadRequest, InternalServerError } from "@curveball/http-errors";
+import { Invites } from "../Services/Email/EmailTemplates"
+
+var ElectionsModel = ServiceLocator.electionsDb();
+var ElectionRollModel = ServiceLocator.electionRollDb();
+var EmailService = ServiceLocator.emailService();
+
+const className = "election.Controllers";
+
+const finalizeElection = async (req: any, res: any, next: any) => {
+    Logger.info(req, `${className}.finalize ${req.election.election_id}`);
+    expectPermission(req.user_auth.roles, permissions.canEditElectionState)
+    if (req.election.state !== 'draft') {
+        var msg = "Election already finalized";
+        Logger.info(req, msg);
+        throw new BadRequest(msg)
+    }
+    var failMsg = "Failed to update Election";
+    req.election.state = 'finalized'
+    const updatedElection = await ElectionsModel.updateElection(req.election, req, `Finalizing election`);
+    if (!updatedElection) {
+        Logger.info(req, failMsg);
+        throw new BadRequest(failMsg)
+    }
+    req.election = updatedElection
+
+    if (req.election.settings.election_roll_type === 'Email') {
+        const electionId = req.election.election_id;
+        const electionRoll = await ElectionRollModel.getRollsByElectionID(electionId, req);
+        if (!electionRoll) {
+            const msg = `Election roll for ${electionId} not found`;
+            Logger.info(req, msg);
+            throw new BadRequest(msg)
```

Comment:
can we move this check above ElectionsModel.updateElection?  Basically want to do all checks, and then have all mutations at the end  (to minimize what could go wrong across different mutations).  If we refactor this to use event-driven-arch later, this point may be moot.

Reply:
Fixed. This was a good idea, wouldn't want someone to forget to add the email roll and end up not being able to fix it.